src/libsyntax_ext/lib.rs

```diff
@@ -150,6 +151,12 @@ pub fn register_builtins(resolver: &mut dyn syntax::ext::base::Resolver,
             SyntaxExtensionKind::LegacyAttr(Box::new(test::expand_bench)), edition
         )
     });
+    register(sym::global_allocator, SyntaxExtension {
+        allow_internal_unstable: Some([sym::rustc_attrs][..].into()),
+        ..SyntaxExtension::default(
+            SyntaxExtensionKind::LegacyAttr(Box::new(global_allocator::expand)), edition
```

Comment:
I think in theory there's no need for this to be a legacy extension, but is this done for other reasons? I could imagine, for example:

* A bit faster since everything is already parsed
* Guaranteed good error messages since spans are always right
* Other builtin attributes still seem to be using legacy attributes

Just wanted to but sure this was intentional!

Reply:
Well, it's legacy because everything built-in is still on legacy (aka AST-based).
We don't have good quasi-quoting that can be used internally to conveniently generate token streams, there are some unresolved issues with error recovery, but the long term plan is to migrate to tokens, yes.

But right now I simply took the existing code from `rustc_allocator` having signature `fn(ast::Item) -> Vec<ast::Item>` and wrapped it into the corresponding AST-based `SyntaxExtensionKind`.

Reply:
Ok that's what I figured, thanks!